google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/sdk/CloudSdkServiceManager.java

```diff
@@ -43,6 +77,203 @@ public void onNewCloudSdkServiceTypeSelected(CloudSdkServiceType newServiceType)
+   *     starting, i.e. deployment or local server.
+   * @param sdkStatusHandler Logger for errors etc.
+   */
+  public void runWhenSdkReady(
+      @Nullable Project project,
+      @NotNull Runnable afterSdkReady,
+      String progressMessage,
+      CloudSdkStatusHandler sdkStatusHandler) {
+    doWait(
+        project,
+        () -> {
+          // at this point the installation should be either ready, failed or user cancelled.
+          // run only if ready.
+          if (CloudSdkService.getInstance().getStatus() == SdkStatus.READY) {
+            ApplicationManager.getApplication().invokeLater(afterSdkReady);
+          }
+        },
+        progressMessage,
+        sdkStatusHandler);
+  }
+
+  /**
+   * Blocks current thread until Cloud SDK is ready for all operations. If process results in error
+   * or user cancel, calls {@link CloudSdkStatusHandler} methods to notify about errors and shows
+   * notifications. This method is expected to be called from non-UI background thread.
+   *
+   * @param project Project to which SDK operation belongs.
+   * @param progressMessage Message to show in progress dialog to identify which process is
+   *     starting, i.e. deployment or local server.
+   * @param sdkStatusHandler Callback to log errors and progress.
+   */
+  public void blockUntilSdkReady(
+      @Nullable Project project, String progressMessage, CloudSdkStatusHandler sdkStatusHandler)
+      throws InterruptedException {
+    CountDownLatch blockingCompletedLatch = new CountDownLatch(1);
+    ApplicationManager.getApplication()
+        .invokeLater(
+            () ->
+                doWait(
+                    project, blockingCompletedLatch::countDown, progressMessage, sdkStatusHandler));
+    blockingCompletedLatch.await();
+  }
+
+  /**
+   * Performs generic wait in a separate thread for SDK to become ready, returns immediately. Must
+   * be called from UI thread.
+   */
+  private void doWait(
+      @Nullable Project project,
+      Runnable afterWaitComplete,
+      String progressMessage,
+      CloudSdkStatusHandler sdkLogging) {
+    CloudSdkService cloudSdkService = CloudSdkService.getInstance();
+    // if not ready, attempt to fix and install now.
+    if (cloudSdkService.getStatus() != SdkStatus.READY && cloudSdkService.isInstallSupported()) {
+      cloudSdkService.install();
+    }
+
+    // no need to wait for install if unsupported or completed.
+    CountDownLatch installationCompletionLatch = new CountDownLatch(isInstallInProgress() ? 1 : 0);
+
+    // listener for SDK updates, waits until install / update is done. uses latch to notify UI
+    // blocking thread.
+    final CloudSdkService.SdkStatusUpdateListener sdkStatusUpdateListener =
+        (sdkService, status) -> {
+          switch (status) {
+            case READY:
+            case INVALID:
+            case NOT_AVAILABLE:
+              installationCompletionLatch.countDown();
+              break;
+            case INSTALLING:
```

Comment:
nit: the style guide suggests to always use a default case. instead of the "INSTALLING" case, you could put in a default that just has the comment "continue waiting for completion" (and the break is redundant here too, no?).

Reply:
good point, thanks!